### ISSUE TITLE
Potential simplification of offline building and caching

### ISSUE THREAD

**Issue by therealbnut** [rust-skia/rust-skia#429]
Hey thanks for rust-skia I use it all the time!

I'm using [sccache](https://github.com/mozilla/sccache), which caches some rust build outputs. In case you haven't encountered it, `ccache` the offical [skia supported way](https://skia.org/user/build) to cache builds and `sccache` adds rust support. We've found that fetching the skia binary on our CI can be fairly flaky, but our `sccache` cache is hosted on with our CI so has much better reliability, and we're using it anyway.

Due to network flakiness we need to separate `cargo fetch` from `cargo build`. On our systems a fetch can be retried, but the build itself is intended to be deterministic.

To separate `fetch` and use `sccache` instead of your skia build caching I believe your recommended steps are:
 * disable fetching  with the environment variables in `build.rs`
 * special-case/duplicate the build.rs fetching of skia alongside our `cargo fetch` command
 * set the environment variables in `build.rs` to specify where we've downloaded skia

I imagine this will all work, but it seems like a lot of setup, only works with un-forked Skia binaries, and also requires us to special-case a bunch of stuff for fetching that adds knowledge of a transitive dependency several layers above our rust code.

---

**TL;DR**

I've had an idea to improve this, and wanted to get your thoughts on it. It might result in less fetching and caching code for you, and might mean you don't need to manually upload binaries with release. The idea is:

You could move `build.rs` (and its module) into `rust-skia/skia`, which becomes a crate (not submodule). Use cargo to do the fetching, as sccache to do the binary caching. You could support both if there's any reason to by putting it behind a `feature` config.

---

The advantage of this will be:
 * A crate means that `cargo fetch` will fetch it, and it will play well with future `rust` ecosystem changes
 * You get the binary caching logic "for free" with `sccache`
 * skia supports `sccache`, so changes to skia itself will be cached and building it will be faster
 * I haven't checked, but you might be able to reuse a Google official ccache repo for Skia if there's a public one
 * If `cargo` has more caching solutions in future using `cargo` in a more vanilla way is more likely to be compatible

The disadvantage (if it's not opt-in) will be:
 * You need to fetching the source even if you only ever need the cached build
 * To get a build binary shared with the community you might need to set up sccache storage
 * If you don't set up a community read-only cache, then people will need all the dependencies needed to build Skia
 * People that want the caching will have to install sccache

However if you make it opt-in, then the only disadvantage I can think of is that the rust source for building skia is split over multiple repos. I'm not sure if that's actually a disadvantage:
 * `rust-skia/skia` for the simple system bindings
 * `rust-skia/skia-safe` for the safe bindings

**Comment by pragmatrix** [comment]
Hello @therealbnut! This is a great idea and I would like follow up on this, but there are a number of roadblocks that may prevent this from happening.

Today I tried to build rust-skia with sccache and ccache on Windows / MSys2, and while the Rust compilation with sccache reduces build times of about 20% for a vanilla build of the workspace, the Skia build refuses to cache results with neither sccache nor ccache. For ccache I found that [ccache doesn't support clang-cl](https://ccache.dev/platform-compiler-language-support.html) and the sccache issue needs to be investigate separately. Then I tried both on macOS and it seems to work fine with the expected faster builds, like I would expect it would do on Linux.

In a first step, we should support caching in Skia at all, this could be done automatically, or using an environment variable, something like `SKIA_CC_WRAPPER=sccache` and document that in the README.

In a second step, we should try to experiment with an additional crate, if possible putting it behind a feature flag, like you suggested, containing all the Skia sources, ninja, and all its dependencies vendored (which [are a lot](https://github.com/rust-skia/skia/blob/4b863f613f8bab8f7e4b0eebec78cd6e5cddb244/DEPS)). Here, one problem is that adding such a huge number of files might make the dependency tracking of cargo, and therefore development based on `cargo check` feedback, too slow. I can remember that this was one of the reasons I had to use an `[include]` section in the `Cargo.toml` of skia-bindings. Though, perhaps this is not an issue anymore, because we are emitting `rerun-if-*` directives from `build.rs`, which should override the list of files being tracked. I will investigate this, too.

All in all I doubt that that solely relying on sccache can replace the prebuilt binaries, for once, because sccache [relies on absolute paths to match](https://github.com/mozilla/sccache#known-caveats), and also the bindings generation is really really slow, but the benefits in other areas are sure worth a try. For example, we could reduce build times with a shared sccache significantly on the CI.

TODO:

- [ ] Support caching
  - [ ] Support setting the `cc_wrapper` argument in the Skia build.
  - [ ] Find out why sccache does not work on Windows to build Skia.
  - [ ] Optimize CI builds
    - [ ] Can we get Azure Blob Storage for free when we use it only for rust-skia?
- [ ] Try to support vendored builds
  - [ ] Introduce a new crate including all the Skia sources and dependencies.
  - [ ] Does it slow down `cargo check`?
  - [ ] Can we publish binaries like ninja.exe in a crate, etc. Or do we need a replacement, or expect ninja to be installed?
  - [ ] Is it larger than 10MB?
  - [ ] Can it put behind a feature `vendored`.

**Comment by therealbnut** [comment]
@pragmatrix thanks for the reply :)

I wonder if the windows caching issues are a Skia bug you could report upstream? I only build on Linux and Mac so I hadn’t noticed the issue (I don’t have access to a windows dev machine).

That’s a good point on the absolute paths, meaning it won’t replace the binary distribution (which also has lower overheads).

You have an interesting insight with `build.rs` and `[include]`, I didn’t even think about that. I’m not sure how it was when you last tried, but apparently `rerun-if-changed` can take a directory now,  which on Mac at least I know should be fast. If you’re relying on sccache then perhaps you can just set it to always rerun, as sccache is designed to efficiently detect what has changed.

Your TODO list looks awesome, thank you for looking into this!
